collector/acceleratos.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type acceleratorsCollector struct {
+	pciDevicesPath string
+	logger         log.Logger
+}
+
+func init() {
+	registerCollector("accelerators", defaultEnabled, NewAcceleratorCollector)
+}
+
+// NewXfrmCollector returns a new Collector exposing XFRM stats.
+func NewAcceleratorCollector(logger log.Logger) (Collector, error) {
+	return &acceleratorsCollector{
+		pciDevicesPath: filepath.Join(*sysPath, "bus/pci/devices"),
+		logger:         logger,
+	}, nil
+}
+
+var (
+	acceleratorCardsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "accelerator", "cards_count"),
+		"number of accelerator cards",
+		[]string{"type"}, nil,
+	)
+
+	// vendor map, add any new vendor to this map
+	vendorToDeviceMap = map[string]map[string]string{
+		// nvidia devices
+		"0x10de": nvidiaDeviceIDsMap,
+	}
+)
+
+func (a *acceleratorsCollector) Update(ch chan<- prometheus.Metric) error {
+	devicesCount := map[string]int{}
+	pciDevices, err := os.ReadDir(a.pciDevicesPath)
+	if err != nil {
+		return fmt.Errorf("failed to read sysfs pci device directory %s: %v", a.pciDevicesPath, err)
+	}
+
+	for _, pciDevice := range pciDevices {
+		vendorID, deviceID, err := a.getVendorAndDeviceIDs(pciDevice)
+		if err != nil {
+			return fmt.Errorf("failed to get vendor and device IDs for device %s: %v", pciDevice.Name(), err)
```

Comment:
Same concern as above regarding errors, is this a common error or is not finding the vendor and device an expected case?

Reply:
The same answer as above: vendor and device files should be present for every pci device on the node. I am guessing that we will see this error only case there something seriously wrong with the node, or because os.ReadFile failed due to some weird error